demo-java/app/src/gms/java/com/example/placesdemo/AutocompleteAddressActivity.java

```diff
@@ -249,4 +293,78 @@ private void clearForm() {
         }
         address1Field.requestFocus();
     }
+
+    // Register the permissions callback, which handles the user's response to the
+    // system permissions dialog. Save the return value, an instance of
+    // ActivityResultLauncher, as an instance variable.
+    private ActivityResultLauncher<String> requestPermissionLauncher =
+            registerForActivityResult(new ActivityResultContracts.RequestPermission(), this::onActivityResult);
+
+    private void checkLocationPermissions() {
+        Toast toast = Toast.makeText(getApplicationContext(), "checking proximity", Toast.LENGTH_SHORT);
+        toast.show();
+        if (ContextCompat.checkSelfPermission(this, ACCESS_FINE_LOCATION)
+                == PackageManager.PERMISSION_GRANTED) {
+            Log.d(TAG, "fine location permission granted");
+            getAndCompareLocations();
+        } else {
+            requestPermissionLauncher.launch(
+                    ACCESS_FINE_LOCATION);
+        }
+    }
+
+    private void onActivityResult(Boolean isGranted) {
+        if (isGranted) {
+            Log.d(TAG, "permission granted upon request");
+            getAndCompareLocations();
+        } else {
+            Log.d(TAG, "permission denied");
+            Toast.makeText(
+                    this,
+                    "User denied location permission. Cannot check proximity to entered address.",
+                    Toast.LENGTH_SHORT)
+                    .show();
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private void getAndCompareLocations() {
+        // TODO: Detect and handle if user has entered or modified the address manually and update
+        // the coordinates variable to the Lat/Lng of the manually entered address. May use
+        // Geocoding API to convert the manually entered address to a Lat/Lng.
+        LatLng enteredLocation = coordinates;
+
+        FusedLocationProviderClient fusedLocationClient =
+                LocationServices.getFusedLocationProviderClient(this);
+
+        fusedLocationClient.getLastLocation()
+                .addOnSuccessListener(this, location -> {
+                    // Got last known location. In some rare situations this can be null.
+                    if (location != null) {
```

Comment:
You can do an early return here to reduce nesting.

i.e.

```
if (location == null) {
    return;
}

// ...
```

Reply:
Thank you!